src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/CodeActions/Razor/WrapAttributesCodeActionProvider.cs

```diff
@@ -0,0 +1,124 @@
+        }
+
+        var syntaxTree = context.CodeDocument.GetSyntaxTree();
+        if (syntaxTree?.Root is null)
+        {
+            return SpecializedTasks.EmptyImmutableArray<RazorVSInternalCodeAction>();
+        }
+
+        var owner = syntaxTree.Root.FindNode(TextSpan.FromBounds(context.StartAbsoluteIndex, context.EndAbsoluteIndex));
+        if (owner is null)
+        {
+            return SpecializedTasks.EmptyImmutableArray<RazorVSInternalCodeAction>();
+        }
+
+        var attributes = FindAttributes(owner);
+        if (attributes.Count == 0)
+        {
+            return SpecializedTasks.EmptyImmutableArray<RazorVSInternalCodeAction>();
+        }
+
+        var first = true;
+        var firstAttributeLine = 0;
+        var indentSize = 0;
+        var sourceText = context.SourceText;
+
+        using var newLinePositions = new PooledArrayBuilder<int>();
+        foreach (var attribute in attributes)
+        {
+            var linePositionSpan = attribute.GetLinePositionSpan(context.CodeDocument.Source);
+
+            if (first)
+            {
+                firstAttributeLine = linePositionSpan.Start.Line;
+                sourceText.TryGetFirstNonWhitespaceOffset(attribute.Span, out var indentSizeOffset);
+                indentSize = linePositionSpan.Start.Character + indentSizeOffset;
+                first = false;
+            }
+            else
+            {
+                if (linePositionSpan.Start.Line != firstAttributeLine)
+                {
+                    return SpecializedTasks.EmptyImmutableArray<RazorVSInternalCodeAction>();
+                }
+
+                if (!sourceText.TryGetFirstNonWhitespaceOffset(attribute.Span, out var startOffset))
+                {
+                    continue;
+                }
+
+                newLinePositions.Add(attribute.SpanStart + startOffset);
+            }
+        }
+
+        if (newLinePositions.Count == 0)
+        {
+            return SpecializedTasks.EmptyImmutableArray<RazorVSInternalCodeAction>();
+        }
+
+        var data = new WrapAttributesCodeActionParams
+        {
+            IndentSize = indentSize,
+            NewLinePositions = newLinePositions.ToArray()
+        };
+
+        var resolutionParams = new RazorCodeActionResolutionParams()
+        {
+            TextDocument = context.Request.TextDocument,
+            Action = LanguageServerConstants.CodeActions.WrapAttributes,
+            Language = RazorLanguageKind.Razor,
+            DelegatedDocumentUri = context.DelegatedDocumentUri,
+            Data = data
+        };
+
+        var action = RazorCodeActionFactory.CreateWrapAttributes(resolutionParams);
+
+        return Task.FromResult<ImmutableArray<RazorVSInternalCodeAction>>([action]);
+    }
+
+    private AspNetCore.Razor.Language.Syntax.SyntaxList<RazorSyntaxNode> FindAttributes(AspNetCore.Razor.Language.Syntax.SyntaxNode owner)
+    {
+        foreach (var node in owner.AncestorsAndSelf())
```

Comment:
>owner.AncestorsAndSelf() [](http://example.com/codeflow?start=29&length=24)

So selection can be anywhere after the start tag name withing the tag? (just a question to make sure I understand the code)